Composer/packages/client/src/store/action/dialog.ts

```diff
@@ -1,159 +1,136 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
-
-import clonedeep from 'lodash/cloneDeep';
-import reject from 'lodash/reject';
-import { DialogInfo } from '@bfc/shared';
-import debounce from 'lodash/debounce';
+import { autofixReferInDialog } from '@bfc/indexers';
 
 import { ActionCreator, State } from '../types';
-import { undoable, Pick } from '../middlewares/undo';
-import luFileStatusStorage from '../../utils/luFileStatusStorage';
+import { undoable } from '../middlewares/undo';
+import { getBaseName } from '../../utils/fileUtil';
 
+import { removeLgFile, createLgFile } from './lg';
 import { ActionTypes } from './../../constants/index';
 import { navTo } from './navigation';
 import { Store } from './../types';
-import httpClient from './../../utils/httpUtil';
-import { setError } from './error';
+import { createLuFile, removeLuFile } from './lu';
+
+interface FileResource {
+  id: string;
+  content: string;
+}
+
+export const removeDialogBase: ActionCreator = async (store, id) => {
+  store.dispatch({
+    type: ActionTypes.REMOVE_DIALOG,
+    payload: { id },
+  });
+};
 
-const pickDialog: Pick = (state: State, args: any[], isStackEmpty) => {
-  const id = args[0];
-  const dialog = state.dialogs.find(dialog => dialog.id === id);
-  let dialogs = clonedeep(state.dialogs);
-  if (!isStackEmpty) {
-    dialogs = reject(dialogs, ['id', id]);
+export const createDialogBase: ActionCreator = async (store, { id, content }) => {
+  const fixedContent = autofixReferInDialog(id, content);
+  const onCreateDialogComplete = store.getState().onCreateDialogComplete;
+  if (typeof onCreateDialogComplete === 'function') {
+    setTimeout(() => onCreateDialogComplete(id));
   }
```

Comment:
dispatch an `onCreateDialogComplete` action at somewhere handles the payload of CREATE_DIALOG

Reply:
I don't know exactly what we do in this onCreateDialogComplete, it seems not aligns with the pattern we are using here, but if possible, it it's just related changes like lg,lu files creation, we can just pre-compute the changes and commit to store once.

Reply:
This is used to create a dialog. we have three actions: createBegin, create and createCancel.  onCreateDialogComplete is used to wait the createBegin complete.
createBegin is used to open 
![image](https://user-images.githubusercontent.com/39758135/78152581-96972800-746c-11ea-8c8f-b853a212b0d5.png)

Reply:
This is required because the extensions do not have direct access to the reducer store, as well as different behavior upon completion of a dialog creation. But the creation flow is async and has disjointed steps. So the initiator of the create dialog flow needs to be able to determine what to do when it is complete.

If there is a better way to model this in the store, I'd be happy to discuss.

Reply:
get it, @a-b-r-o-w-n i talked to Leilei and dig it a little bit, it was a 2 step actions and extension needs to get the result inorder to further update the action node, i don't have a good proposal here, maybe we can do better by having a more generic approach for extensions to get notification instead of directly inject functions into store, but that's would reduce this essentially complex, and will likely be a overkill. 

let's leave it for now, as far as in this PR, the goal is decouple the concern of persistence from the upper layer (the editors and other components that emit actions), which they can assume the store is always available and update correctly and fast.